### ISSUE TITLE
How get similar items using model.tem_embeddings???

### ISSUE THREAD

**Issue by hammadkhann** [lyst/lightfm#353]
I want to get similar items by using item_embeddings can you guide me how can I do that using cosine_similarity ??

**Comment by fischjer4** [comment]
`def similar_items(internal_idx, item_feats_mtx, model, N=10):
    item_representations = item_feats_mtx.dot(model.item_embeddings)

    # Cosine similarity
    scores = item_representations.dot(item_representations[internal_idx, :])
    item_norms = np.linalg.norm(item_representations, axis=1)
    scores /= item_norms

    best = np.argpartition(scores, -N)[-N:]
    return sorted(zip(best, scores[best] / item_norms[internal_idx]),
                  key=lambda x: -x[1])`

Hi @hammadkhann, I found this on the internet and it works great for me. `internal_idx` is the item's index in the matrix which you can get from `dataset.mapping()[2][<your item ID>]`.  Hopefully, someone with more experience will comment on what exactly is happening behind the scene because I am still a little bit confused as well. But, I wanted to post this to at least get you started.

**Comment by maciejkula** [comment]
This is correct, but please use the [get_item/user_representation](http://lyst.github.io/lightfm/docs/lightfm.html#lightfm.LightFM.get_item_representations) methods to get the representation matrices.

**Comment by fischjer4** [comment]
@maciejkula Just to clarify, you're suggesting to simply replace the first line 
`item_representations = item_feats_mtx.dot(model.item_embeddings)` 
with 
`item_representations = model.get_item_representations()[1]` correct?

**Comment by hammadkhann** [comment]
Great

**Comment by hammadkhann** [comment]
Can I filter my item representation based on certain filter criteria for e.g city,country and category and then find the top similar items??

**Comment by hammadkhann** [comment]
Or should I put my city,country and category as feature in my training data and then find the simple cosine similarity as above?

**Comment by fischjer4** [comment]
@hammadkhann Based off of your question and answer from [#343](https://github.com/lyst/lightfm/issues/343)  my guess is that you would filter the `item_representations` to only include the representations of the items which meet your criteria.
assume `indices_that_meet` is an array of item indices that meet your criteria.
```
filtered_item_reps = item_representations[indices_that_meet, ] 
```
and then use `filtered_item_reps` instead of `item_representations` in the`similar_items` function

**Comment by hammadkhann** [comment]
Thanks let me try @fischjer4

**Comment by hammadkhann** [comment]
I am confused about filters you define like I just define my filters as [city = "Karachi", country="Pakistan"]??

**Comment by hammadkhann** [comment]
@maciejkula @fischjer4   what do you guys  prefer filtering item representation space and using cosine similarity to get similar items and pass those item_ids to predict function or I filter items using item data and then pass filtered item_ids to the predict function on real time prediction? I am talking in terms of speed and memory usage or is there any other optimal way to filter items on real time and get recommendations?

**Comment by fischjer4** [comment]
@hammadkhann regarding the filters, the `filters` is a list of the feature _labels_ that you used when fitting your dataset (`dataset.fit(..., item_features=items_feat_labels, ...)`). So if the label was simply "Karachi" and "Pakistan", then your `filters = ["Karachi", "Pakistan"]`

**Comment by hammadkhann** [comment]
@fischjer4 thanks

**Comment by hammadkhann** [comment]
@fischjer4  I am getting only two items in filtered_item_feat_idxs when used filter as ["Karachi","Pakistan"] although there are many items containing these filters :/

**Comment by fischjer4** [comment]
@hammadkhann You should only get two in `filtered_item_feat_idxs ` if you only supply it two filters. What `filtered_item_feat_idxs` represents is the mapping from the external feature name to the internal index that that item feature is in the matrix. The matrix has no clue what "Pakistan" is, so you map "Pakistan" to its index in the matrix, then use that index to find items that meet by calling `find_filtered_item_idxs ` 
Make sense?

**Comment by fischjer4** [comment]
Are ["Karachi","Pakistan"] the actual feature labels that you used to fit the dataset? What does `filtered_item_feat_idxs` contain, are they legit indices? When you built your `item_feat_mtx` did you pass it your item features, specifically were their items that had "Karachi" and "Pakistan" as features in your call to `dataset.build_item_features()`

**Comment by hammadkhann** [comment]
![image](https://user-images.githubusercontent.com/22725956/43921255-9b5e6b58-9c34-11e8-8d4f-bf699b078dc5.png)
This is the feature list i am passing when building item_features

**Comment by hammadkhann** [comment]
@fischjer4  If I am filtering with the above tags like Fast Food its getting me the indices but for Karachi and Pakistan its returning null

**Comment by hammadkhann** [comment]
@fischjer4  I fixed that issue kindly tell me can I do and operations between the filters means only those indices will come which have combination of "Karachi" and "Pakistan".??

**Comment by SantiDu** [comment]
> @maciejkula Just to clarify, you're suggesting to simply replace the first line
> `item_representations = item_feats_mtx.dot(model.item_embeddings)`
> with
> `item_representations = model.get_item_representations()[1]` correct?

I think he means this:
`item_representations = item_feats_mtx.dot(model.get_user_representations()[1])`
